wagtail_streamfield_migration_toolkit/test/tests/test_bad_data.py

```diff
@@ -0,0 +1,343 @@
+            }
+        )
+        stream_block = instance.content.stream_block
+        instance.content = StreamValue(
+            stream_block=stream_block, stream_data=raw_data, is_lazy=True
+        )
+        instance.save()
+
+        revision = instance.save_revision()
+        revision.created_at = timezone.now()
+        revision.save()
+
+        raw_data = instance.content.raw_data
+        raw_data = raw_data[:2]
+        stream_block = instance.content.stream_block
+        instance.content = StreamValue(
+            stream_block=stream_block, stream_data=raw_data, is_lazy=True
+        )
+        instance.save()
+
+    def test_migrate(self):
+        with self.assertRaisesMessage(
+            InvalidBlockDefError, "No current block def named invalid_name1"
+        ):
+            self.apply_migration(
+                operations_and_block_path=[
+                    (
+                        RenameStructChildrenOperation(
+                            old_name="char1", new_name="renamed1"
+                        ),
+                        "invalid_name1",
+                    )
+                ],
+                revisions_from=None,
+            )
+
+
+class TestExceptionRaisedForLiveRevision(TestCase, MigrationTestMixin):
+    """Exception should always be raised when applying migration if it occurs while migrating the
+    live revision data"""
+
+    model = models.SamplePage
+
+    @classmethod
+    def setUpTestData(cls):
+        instance = factories.SamplePageFactory(
+            content__0__char1__value="Char Block 1",
+            content__1="nestedstruct",
+        )
+
+        raw_data = instance.content.raw_data
+        raw_data.append(
+            {
+                "type": "invalid_name1",
+                "id": "0001",
+                "value": {"char1": "foo", "char2": "foo"},
+            }
+        )
+        raw_data.append(
+            {
+                "type": "invalid_name1",
+                "id": "0002",
+                "value": {"char1": "foo", "char2": "foo"},
+            }
+        )
+        stream_block = instance.content.stream_block
+        instance.content = StreamValue(
+            stream_block=stream_block, stream_data=raw_data, is_lazy=True
+        )
+        instance.save()
+
+        revision = instance.save_revision()
+        revision.created_at = timezone.now() - datetime.timedelta(days=(5))
+        revision.save()
+
+        raw_data = instance.content.raw_data
+        raw_data = raw_data[:2]
+        stream_block = instance.content.stream_block
+        instance.content = StreamValue(
+            stream_block=stream_block, stream_data=raw_data, is_lazy=True
+        )
+        instance.live_revision = revision
+        instance.save()
+
+        for i in range(1, 5):
+            revision = instance.save_revision()
+            revision.created_at = timezone.now() - datetime.timedelta(days=(5 - i))
+            revision.save()
```

Comment:
It looks like we have some duplicated code here too. Perhaps  `RevisionMigrationTestCase` and `BadDataMigrationTestCase` classes are in order.